pkg/profiler/buffer.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package profiler
+
+import "encoding/binary"
+
+type EfficientBuffer []byte
+
+// Slice returns a slice re-sliced from the original EfficientBuffer.
+// This is useful to efficiently write byte by byte, for example, when
+// setting BPF maps without incurring in extra allocations in the writing
+// methods, or changing the capacity of the underlying memory buffer.
+//
+// Callers are responsible to ensure that there is enough capacity left
+// for the passed size.
+func (eb *EfficientBuffer) Slice(size int) EfficientBuffer {
+	newSize := len(*eb) + size
+	subSlice := (*eb)[len(*eb):newSize]
+	// Extend its length.
+	*eb = (*eb)[:newSize]
+	return subSlice
+}
+
+func (eb *EfficientBuffer) PutUint64(v uint64) {
+	binary.LittleEndian.PutUint64((*eb)[:8], v)
+	*eb = (*eb)[8:]
```

Comment:
This is clever, but different from what people are likely used to from the `binary` package. I would suggest to write a function comment that notes that the function writes the uint64 in little endian _and_ advances the buffer.